lib/CL/pocl_util.c

```diff
@@ -2124,3 +2126,36 @@ bitcode_is_spirv_kernel (const char *bitcode, size_t size)
 
   return is_opencl;
 }
+
+int
+bitcode_is_vulkan_spirv (const char *bitcode, size_t size)
+{
+  const uint32_t *bc32 = (const uint32_t *)bitcode;
+  unsigned location = 0;
+  uint32_t header_magic = htole32 (bc32[location++]);
+
+  if ((size < 20) || (header_magic != SPIRV_MAGIC))
+    return 0;
+
+  // skip version, generator, bound, schema
+  location += 4;
+  int is_shader = 0;
+  uint32_t instruction, value;
+  do
+    {
+      instruction = htole32 (bc32[location++]);
+      value = htole32 (bc32[location++]);
+      if (value == ShaderExecModel)
+        is_shader = 1;
+    }
+  while (instruction == OpCapab);
+
+  /* SPIR-V but not OpenCL-type. */
+  if (!is_shader)
+    {
+      POCL_MSG_ERR ("SPIR-V binary provided, but is not using Shader mode."
```

Comment:
Please document this aspect in the user manual. Does this mean we need to input Vulkan Compute shader SPIR-V:s for the Vulkan driver to work, the OpenCL profile ones do not work?

Reply:
That's indeed the case. The Sylkan paper I mentioned above [has some details about the semantic differences](https://dl.acm.org/doi/fullHtml/10.1145/3456669.3456683#sec-7).

Reply:
Better document this in the docs too.

Reply:
documented in `opencl_status.rst`